epics/cache.js

```diff
@@ -0,0 +1,100 @@
+/**
+ * Provides a cache outside of Redux state that can optimistically update state
+ * before an asynchronous API call returns
+ *
+ * @module CacheMiddleware
+ */
+import Rx from 'rxjs';
+import { combineEpics } from 'redux-observable';
+import {
+	cacheSuccess,
+} from '../actions/cacheActionCreators';
+
+import {
+	makeCache,
+	cacheReader,
+	cacheWriter,
+} from '../util/cacheUtils';
+
+
+export function checkEnable() {
+	if (typeof window !== 'undefined' && window.location) {
+		const params = new URLSearchParams(window.location.search.slice(1));
+		return !params.has('__nocache');
+	}
+	return true;
+}
+
+/**
+ * The cache middleware triggers a 'set'/store action when new data is received
+ * from the API (API_SUCCESS), and is queried when queries are sent to the API
+ * (API_REQUEST). These events trigger cache-specific events, CACHE_SET and
+ * CACHE_QUERY, which are then used to update the cache or update the
+ * application state (CACHE_SUCCESS)
+ *
+ * @returns {Function} the curried state => action => next middleware function
+ */
+
+// get a cache, any cache (that conforms to the Promise-based API)
+const cache = makeCache();
+
+// get a function that can read from the cache for a specific query
+const readCache = cacheReader(cache);
+// get a function that can write to the cache for a specific query-response
+const writeCache = cacheWriter(cache);
+
+/**
+ * Listen for any action that should clear cached state
+ *
+ * Note that this will clear the cache without emitting an action
+ */
+export const cacheClearEpic = action$ =>
+	action$.ofType('LOGOUT_REQUEST', 'CACHE_CLEAR')
+		.flatMap(() => cache.clear())  // wait for cache to clear before continuing
+		.flatMap(() => Rx.Observable.empty());
+
+/**
+ * Listen for any action that should set cached state with a
+ * `{ queries, responses }` payload
+ *
+ * API_SUCCESS means there is fresh data ready to be stored - extract the
+ * queries and their responses, then dispatch `CACHE_SET` actions with each
+ * pair
+ *
+ * Not that this will set the cache without emitting an action
+ */
+export const cacheSetEpic = action$ =>
+	action$.ofType('API_SUCCESS', 'CACHE_SET')
```

Comment:
For these cache epics, you mentioned "...without emitting an action". Does that mean that the incoming actions in `action$` never reach the reducer? Or do they and the reducers are now a no-op? Just curious of the order of operations with these actions + epic middleware + redux reducer.